File: sdks/go/pkg/beam/core/util/reflectx/util_test.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package reflectx

import (
	"fmt"
	"reflect"
	"testing"
)

type foo struct {
	N int
}

func (f *foo) String() string {
	return fmt.Sprintf("&foo{%v}", f.N)
}

func TestMergeMaps(t *testing.T) {
	tests := []struct {
		a, b, exp interface{}
	}{
		{
			a:   map[string]*foo{"a": {1}},
			b:   map[string]*foo{"a": {2}},
			exp: map[string]*foo{"a": {2}},
		},
		{
			a:   map[string]*foo{"a": {1}},
			b:   nil,
			exp: map[string]*foo{"a": {1}},
		},
		{
			a:   map[string]*foo{"a": {1}},
			b:   map[string]*foo{"b": {1}},
			exp: map[string]*foo{"a": {1}, "b": {1}},
		},
		{
			a:   map[string]*foo{"a": {1}},
			b:   map[string]*foo{"a": nil, "b": {1}},
			exp: map[string]*foo{"b": {1}},
		},
		{
			a:   map[string]*foo{"a": {1}},
			b:   map[string]*foo{"a": {2}, "not_present": nil},
			exp: map[string]*foo{"a": {2}},
		},
	}

	for _, test := range tests {
		orig := fmt.Sprintf("%v", test.a) // print before mutation

		UpdateMap(test.a, test.b)
		if !reflect.DeepEqual(test.a, test.exp) {
			t.Errorf("UpdateMap(%v,%v) = %v, want %v", orig, test.b, test.a, test.exp)
		}
	}
}

func TestShallowClone(t *testing.T) {
	tests := []interface{}{
		nil,
		2,
		foo{4},
		&foo{23},
		[]string{},
		[]string{"a", "c", "e"},
		map[string]string{},
		map[string]string{"a": "c", "e": "g"},
	}
	for _, test := range tests {
		actual := ShallowClone(test)
		if !reflect.DeepEqual(actual, test) {
			t.Errorf("ShallowClone(%v) = %v, want id", test, actual)
		}
	}
}

func TestShallowCloneNil(t *testing.T) {
	var a []string

	ac := ShallowClone(a)
	if !reflect.DeepEqual(ac, a) {
		t.Errorf("ShallowClone(%v) = %v, want id", a, ac)
	}

	var b map[string]string

	bc := ShallowClone(b)
	if !reflect.DeepEqual(bc, b) {
		t.Errorf("ShallowClone(%v) = %v, want id", b, bc)
	}
}

File: sdks/python/apache_beam/io/gcp/datastore/v1new/util_test.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""Tests for util.py."""
# pytype: skip-file

from __future__ import absolute_import

import unittest

from apache_beam.io.gcp.datastore.v1new import util


class MovingSumTest(unittest.TestCase):

  TIMESTAMP = 1500000000

  def test_bad_bucket_size(self):
    with self.assertRaises(ValueError):
      _ = util.MovingSum(1, 0)

  def test_bad_window_size(self):
    with self.assertRaises(ValueError):
      _ = util.MovingSum(1, 2)

  def test_no_data(self):
    ms = util.MovingSum(10, 1)
    self.assertEqual(0, ms.sum(MovingSumTest.TIMESTAMP))
    self.assertEqual(0, ms.count(MovingSumTest.TIMESTAMP))
    self.assertFalse(ms.has_data(MovingSumTest.TIMESTAMP))

  def test_one_data_point(self):
    ms = util.MovingSum(10, 1)
    ms.add(MovingSumTest.TIMESTAMP, 5)
    self.assertEqual(5, ms.sum(MovingSumTest.TIMESTAMP))
    self.assertEqual(1, ms.count(MovingSumTest.TIMESTAMP))
    self.assertTrue(ms.has_data(MovingSumTest.TIMESTAMP))

  def test_aggregates_within_window(self):
    ms = util.MovingSum(10, 1)
    ms.add(MovingSumTest.TIMESTAMP, 5)
    ms.add(MovingSumTest.TIMESTAMP + 1, 3)
    ms.add(MovingSumTest.TIMESTAMP + 2, 7)
    self.assertEqual(15, ms.sum(MovingSumTest.TIMESTAMP + 3))
    self.assertEqual(3, ms.count(MovingSumTest.TIMESTAMP + 3))

  def test_data_expires_from_moving_window(self):
    ms = util.MovingSum(5, 1)
    ms.add(MovingSumTest.TIMESTAMP, 5)
    ms.add(MovingSumTest.TIMESTAMP + 3, 3)
    ms.add(MovingSumTest.TIMESTAMP + 6, 7)
    self.assertEqual(10, ms.sum(MovingSumTest.TIMESTAMP + 7))
    self.assertEqual(2, ms.count(MovingSumTest.TIMESTAMP + 7))


class DynamicWriteBatcherTest(unittest.TestCase):
  def setUp(self):
    self._batcher = util.DynamicBatchSizer()

  # If possible, keep these test cases aligned with the Java test cases in
  # DatastoreV1Test.java
  def test_no_data(self):
    self.assertEqual(
        util.WRITE_BATCH_INITIAL_SIZE, self._batcher.get_batch_size(0))

  def test_fast_queries(self):
    self._batcher.report_latency(0, 1000, 200)
    self._batcher.report_latency(0, 1000, 200)
    self.assertEqual(util.WRITE_BATCH_MAX_SIZE, self._batcher.get_batch_size(0))

  def test_slow_queries(self):
    self._batcher.report_latency(0, 10000, 200)
    self._batcher.report_latency(0, 10000, 200)
    self.assertEqual(100, self._batcher.get_batch_size(0))

  def test_size_not_below_minimum(self):
    self._batcher.report_latency(0, 30000, 50)
    self._batcher.report_latency(0, 30000, 50)
    self.assertEqual(util.WRITE_BATCH_MIN_SIZE, self._batcher.get_batch_size(0))

  def test_sliding_window(self):
    self._batcher.report_latency(0, 30000, 50)
    self._batcher.report_latency(50000, 5000, 200)
    self._batcher.report_latency(100000, 5000, 200)
    self.assertEqual(200, self._batcher.get_batch_size(150000))


if __name__ == '__main__':
  unittest.main()
